File: func.py
#!/usr/bin/env python

def parse_label_time(timestamp):
    # This is a horrible idea.
    secs, msecs = timestamp.split('.', 1)
    return int(secs) * 1000 + int(msecs[:3])

def parse_labels(labels):
    for line in labels:
        line = line.strip()
        if not line:
            continue

        if line.startswith('#'):
            continue

        start, stop, message = line.split(None, 2)
        start = parse_label_time(start)
        stop = parse_label_time(stop)
        yield start, stop, message

def sort_messages(timeline):
    times = { 0: [] }
    for label in timeline:
        start, stop, message = label
        times.setdefault(start, []).append(('add', label))
        times.setdefault(stop, []).append(('remove', label))

    active_labels = []
    for time, operations in sorted(times.iteritems()):
        for operation in operations:
            op, label = operation
            if op == 'add':
                active_labels.append(label)
            elif op == 'remove':
                active_labels.remove(label)

        yield time, [message for (start, stop, message) in active_labels]

def srt_timecode(msecs):
    value, ms = divmod(msecs, 1000)
    value, s = divmod(value, 60)
    h, m = divmod(value, 60)
    return '%02d:%02d:%02d,%03d' % (h, m, s, ms)

def make_srt(sorted_messages):
    srt_chunks = []
    i = 1
    sorted_messages = list(sorted_messages)
    for idx, (time, messages) in enumerate(sorted_messages):
        if not messages:
            continue

        start = time
        stop = sorted_messages[idx+1][0]

        srt_chunks.append("""
%d
%s --> %s
%s
""" % (i, srt_timecode(start), srt_timecode(stop), '\n'.join(messages)))
        i += 1

    return ' '.join(srt_chunks)

def labels2srt(labels):
    parsed_labels = parse_labels(labels)
    sorted_messages = sort_messages(parsed_labels)
    srt = make_srt(sorted_messages)
    return srt

File: func.js
(function(exports) {

    function parseLabelTime(timestamp) {
        // This is a horrible idea.
        var parsed = timestamp.split('.', 2);
        var secs = parsed[0], msecs = parsed[1];
        return parseInt(secs, 10) * 1000 + parseInt(msecs.slice(0, 3), 10);
    }

    function parseLabels(lines) {
        var labels = [];

        lines.forEach(function(line) {
            line = line.trim();

            if (!line)
                return;

            if (line.startsWith('#'))
                return;

            var parsed = line.match(/(.*?)\s+(.*?)\s+(.*)/);
            var start = parseLabelTime(parsed[1]);
            var stop = parseLabelTime(parsed[2]);
            var message = parsed[3];
            labels.push({ start: start, stop: stop, message: message });
        });

        return labels;
    }

    function srtTimecode(msecs) {
        var ms = Math.floor(msecs % 1000);
        var value = msecs / 1000;

        var s = Math.floor(value % 60);
        value = value / 60;

        var m = Math.floor(value % 60);
        var h = Math.floor(value / 60);

        function pad(n, z) {
            n = '' + n;
            while (n.length < z)
                n = '0' + n;
            return n;
        }

        return (pad(h, 2) + ':' + pad(m, 2) + ':' + pad(s, 2) + ',' + pad(ms, 3));
    }

    function formatSrtChunk(label, i) {
        var lines = [];

        lines.push('');
        lines.push(i + 1);
        lines.push(srtTimecode(label.start) + ' --> ' + srtTimecode(label.stop));
        lines.push(label.message);

        return lines.join('\n');
    }

    function labelsToSrt(labels) {
        var lines = labels.split('\n');
        var srt = parseLabels(lines).map(formatSrtChunk).join('\n');
        return srt;
    }

    exports.labelsToSrt = labelsToSrt;

})(window);
